### PR TITLE
Tutorial for setting up AppFlowy on Hetzner Cloud

### PR DESCRIPTION
**Pull request by suntorytimed**
* using Hetzner Cloud server
* using Hetzner Cloud Object Storage

The server and credentials used in the tutorial and shown in the screenshot got deleted and don't need to be censored.

I have read and understood the Contributor's Certificate of Origin available at the end of 
https://raw.githubusercontent.com/hetzneronline/community-content/master/tutorial-template.md
and I hereby certify that I meet the contribution criteria described in it.
Signed-off-by: Stefan Weiberg <github@hibiki.eu>

### CODE REVIEW

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+author_description: ""
+language: "en"
+available_languages: ["en"]
+header_img: "header-4"
+cta: "objectstorage"
+---
+
+## Introduction
+
+In this tutorial we are going to setup our own AppFlowy Cloud server using the official documentation by AppFlowy. We will use a Hetzner Cloud Server and the new Hetzner Object Storage. This tutorial won't cover how you create a new project called "AppFlowy", and how Hetzner Cloud servers are created. We will purely focus on setting up AppFlowy, setting up a bucket and connecting it to our AppFlowy Cloud server.
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:21]: Maybe add a few links here?
```suggestion
In this tutorial we are going to setup our own [AppFlowy](https://appflowy.io/) Cloud server using the official documentation by AppFlowy. We will use a [Hetzner Cloud Server](https://hetzner.cloud) and the new [Hetzner Object Storage](https://docs.hetzner.com/storage/object-storage). This tutorial won't cover how you create a new project called "AppFlowy", and how Hetzner Cloud servers are created. We will purely focus on setting up AppFlowy, setting up a bucket and connecting it to our AppFlowy Cloud server.
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+# configure your admin account
+GOTRUE_ADMIN_EMAIL=admin@example.com
+GOTRUE_ADMIN_PASSWORD=password
+
+# input your (sub-)domain, which you configured in the Hetzner DNS service
+API_EXTERNAL_URL=http://your-host
+
+```
+
+Remove following lines from the deploy.env, as we won't use those services:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:151]:
```suggestion
Remove following lines from the `deploy.env`file, as we won't use those services:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+Don't forget to remove the minio_data volume from the volumes at the end of the docker-compose.yml. Only the postgres_data volume should remain:
+
+```
+volumes:
+  postgres_data:
+```
+
+### Step 2.4 - Adjust the deploy.env
+
+Now we need to adjust the deploy.env file to use our own passwords and point to the Hetzner Object Storage. I recommend to keep the order of values identical, as we are going to commit our changes locally. This makes rebasing easier. At the time of creation of this tutorial you should adjust following values with your own password and SMTP configuration:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:119]:
```suggestion
Now we need to adjust the `deploy.env` file to use our own passwords and point to the Hetzner Object Storage. I recommend to keep the order of values identical, as we are going to commit our changes locally. This makes rebasing easier. At the time of creation of this tutorial you should adjust following values with your own password and SMTP configuration:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+APPFLOWY_S3_MINIO_URL=https://appflowy-5d3f1744a2494fb5b49c983e3eff3e17.fsn1.your-objectstorage.com # this is where the previously copied Object Storage URL is pasted
+APPFLOWY_S3_ACCESS_KEY=<your access key> # paste your previously generated access key from the S3 credentials
+APPFLOWY_S3_SECRET_KEY=<your secret key> # paste your previously generated secret key from the S3 credentials
+APPFLOWY_S3_BUCKET=appflowy # this isn't really the name of your bucket, but the root folder in your bucket
+#APPFLOWY_S3_REGION=us-east-1
+```
+
+### Step 2.5 - Add and commit your changes
+
+Now we use git to store our changes in the git log and make it easier to pull updates from upstream. By using this method every change of defaults in the deploy.env and docker-compose.yml will result in merge conflicts when pulling an update, which you need to resolve before restarting/updating your server.
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:181]:
```suggestion
Now we use git to store our changes in the git log and make it easier to pull updates from upstream. By using this method every change of defaults in the `deploy.env` and `docker-compose.yml` will result in merge conflicts when pulling an update, which you need to resolve before restarting/updating your server.
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+Clone the AppFlowy-Cloud repository to your desired location and `cd` into the cloned repository:
+
+```
+git clone https://github.com/AppFlowy-IO/AppFlowy-Cloud
+cd AppFlowy-Cloud
+```
+
+### Step 2.2 - Setup the nginx proxy of AppFlowy Cloud to use your SSL certificate
+
+Put your SSL certificate and private key in the folder ./nginx/ssl/ or adjust the location in the docker-compose.yml to point to your own certificates. Alternatively you can create a Let's Encrypt SSL certificate via DNS challenge on the Hetzner DNS service by using lego:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:70]:
```suggestion
Put your SSL certificate and private key in the folder `./nginx/ssl/` or adjust the location in the `docker-compose.yml` to point to your own certificates. Alternatively you can create a Let's Encrypt SSL certificate via DNS challenge on the Hetzner DNS service by using lego:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+```
+go install github.com/go-acme/lego/v4/cmd/lego@latest
+~/go/bin/lego --domains="appflowy.example.com" --email="admin@example.com" --http run
+```
+
+And adjust the docker-compose.yml to point to the output folder of lego:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:77]:
```suggestion
And adjust the `docker-compose.yml` to point to the output folder of lego:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+      - ${NGINX_TLS_PORT:-443}:443
+    volumes:
+      - ./nginx/nginx.conf:/etc/nginx/nginx.conf
+      - ./.lego/certificates/appflowy.example.com.crt:/etc/nginx/ssl/certificate.crt
+      - ./.lego/certificates/appflowy.example.com.key:/etc/nginx/ssl/private_key.key
+```
+
+### Step 2.3 - Remove the minio service from the docker-compose.yml
+
+Remove the minio service from the docker-compose.yml by deleting following lines:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:94]:
```suggestion
Remove the minio service from the `docker-compose.yml` by deleting following lines:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    environment:
+      - MINIO_BROWSER_REDIRECT_URL=http://localhost/minio
+      - MINIO_ROOT_USER=${APPFLOWY_S3_ACCESS_KEY:-minioadmin}
+      - MINIO_ROOT_PASSWORD=${APPFLOWY_S3_SECRET_KEY:-minioadmin}
+    command: server /data --console-address ":9001"
+    volumes:
+      - minio_data:/data
+```
+
+Don't forget to remove the minio_data volume from the volumes at the end of the docker-compose.yml. Only the postgres_data volume should remain:
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:110]:
```suggestion
Don't forget to remove the `minio_data` volume from the volumes at the end of the `docker-compose.yml`. Only the `postgres_data `volume should remain:
```

File: tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+Now we use git to store our changes in the git log and make it easier to pull updates from upstream. By using this method every change of defaults in the deploy.env and docker-compose.yml will result in merge conflicts when pulling an update, which you need to resolve before restarting/updating your server.
+
+```
+git add deploy.env docker-compose.yml
+git commit -m "my own AppFlowy deployment on Hetzner Cloud"
+```
+
+## Step 3
+
+Start you AppFlowy Cloud server via docker-compose (or docker compose, depending on your installation method):
```
Comment (LenzGr) [inline tutorials/install-and-configure-appflowy-with-hetzner-object-storage/01.en.md:190]:
```suggestion
Start you AppFlowy Cloud server via `docker-compose` (or `docker compose`, depending on your installation method):
```

**Comment by svenja11** [comment]
Thank you @suntorytimed! I reviewed your tutorial and I made some changes.

The [first commit](https://github.com/hetzneronline/community-content/pull/959/commits/2fc8edae237ba6314db692850fe67bc0e091025e) are mainly just spelling and formatting updates. I also removed "Hetzner cloud server" because we try to keep the tutorials general and you could install AppFlowy Cloud on any server if you wanted to.

The [second commit](https://github.com/hetzneronline/community-content/pull/959/commits/6c88f3009f7b283fb0554a6a8cbf670251fa3eb9) has all the big changes:

* **Step 1**
  
  If there already are existing tutorials or guides for something, we usually just link to them in the community tutorial instead of explaining it again. This makes it easier to keep everything up-to-date and we don't need to check several tutorials or guides in case anything changes. In this tutorial, this is the case with "Creating a Bucket" and "Generating S3 credentials". I removed those explanations with their screenshots and linked to the existing getting starteds instead.

* **Step 2**
  
  I wasn't sure if it was clear enough so I added a little bit more information.

* **Step 3**
  
  If the password in `docker-compose.yml` isn't identical to the password in `deploy.env`, you won't be able to login. I added a brief section in `docker-compose.yml` that mentions the password and added a respective note in `deploy.env` too.

  If you don't add a key to the ai service in the docker compose file, the ai service container will keep restarting over and over again. So added a step to either add a key or comment out the service.

* **Step 4**
  
  You mentioned that they should change the password everywhere it's mentioned but it's actually enough to update the password variable at the beginning of the file. So I changed that.

  I also added a little note why you should set up SMTP

* **Step 6**
  
  When I followed your tutorial, I was struggling with this a bit. I added the steps that I missed and that I think might be helpful to others.

  It also threw me off that I suddenly wasn't able to log in anymore after restart. It gave me error messages and I thought I must have done something wrong. Once you follow the magic link, it's back to normal again. I added this info to avoid confusion for others.

Please review my changes and let me know if you're not fine with any of the changes I made. You can also make your own changes again.

**Comment by suntorytimed** [comment]
@svenja11 thanks for your review and all the improvements!

```
Open `docker-compose.yml` with you favourite text editor and make the following changes:

* Edit the postgres service

  Change the default value `password` to a secure password (in this example `securepassword`).

  ```yml
    postgres:
      restart: on-failure
      image: pgvector/pgvector:pg16
      environment:
        - POSTGRES_USER=${POSTGRES_USER:-postgres}
        - POSTGRES_DB=${POSTGRES_DB:-postgres}
        - POSTGRES_PASSWORD=${POSTGRES_PASSWORD:-securepassword}
        - POSTGRES_HOST=${POSTGRES_HOST:-postgres}
  ```
```
This step is actually not required, as the password will be set in the deploy.env file as `POSTGRES_PASSWORD`. Therefore
```
  # replace the value of ${POSTGRES_PASSWORD} with
  # the password you set in docker-compose.yml
  # change the secret
```
is incorrect, as what you set in the deploy.env for ${POSTGRES_PASSWORD} overrides the value in the docker-compose.yml and not the other way around. Afaik. the password in the URLs need to be changed, as they are not pulled from an environment variable.

**Comment by svenja11** [comment]
Now I'm confused. 😅  
I tested it several times and this was the only way I could get it to work. If the password wasn't the same in both files, it gave me the error message "password mismatch".  
But I just tested it again and now it works as you said. 🤷

I just checked and they did some changes to the password in [this quite recent MR](https://github.com/AppFlowy-IO/AppFlowy-Cloud/pull/901/files). Maybe this broke something and they were able to fix it in the meantime.

Either way, I will remove this section as you suggested.👍

**Comment by suntorytimed** [comment]
Yes, I see it now. The AppFlowy-Cloud project is  a very active one and still in an early beta stage. That's why they sometimes change quite a lot in their deployment.

**Pull request merged**